exist-core/src/main/java/org/exist/xquery/functions/fn/FunXmlToJson.java

```diff
@@ -0,0 +1,246 @@
+                    new SequenceType[]{
+                            new FunctionParameterSequenceType("node", Type.NODE,
+                                    Cardinality.ZERO_OR_ONE, "The input node"),
+                            new FunctionParameterSequenceType("options", Type.MAP,
+                                    Cardinality.ONE, "The options map")
+                    },
+                    new FunctionReturnSequenceType(Type.STRING, Cardinality.ZERO_OR_ONE,
+                            "The JSON representation of the input node")
+            )
+    };
+
+    public FunXmlToJson(XQueryContext context, FunctionSignature signature) { super(context, signature); }
+
+    public Sequence eval(Sequence[] args, Sequence contextSequence) throws XPathException {
+        Sequence result;
+        final Sequence seq = (getArgumentCount() > 0) ? args[0] : Sequence.EMPTY_SEQUENCE;
+        //TODO: implement handling of options
+        final MapType options = (getArgumentCount() > 1) ? (MapType) args[1].itemAt(0) : new MapType(context);
+
+        if (seq.isEmpty())
+            {result = Sequence.EMPTY_SEQUENCE;}
+        else {
+            result = new ValueSequence();
+            final Item item = seq.itemAt(0);
+            if (item.getType() != Type.DOCUMENT && item.getType() != Type.ELEMENT) {
+                throw new XPathException(ErrorCodes.FOJS0006, "Invalid XML representation of JSON.");
+            }
+            final NodeValue nodeValue = (NodeValue) item;
+            final StringWriter stringWriter = new StringWriter();
+            nodeValueToJson(nodeValue, stringWriter);
+            final String jsonString = stringWriter.toString();
+            result.add(new StringValue(jsonString));
+        }
+        return result;
+    }
+
+    /**
+     * Generate a JSON representation of a NodeValue which conforms to
+     * https://www.w3.org/TR/xpath-functions-31/schema-for-json.xsd
+     * Traverse a NodeValue via XMLStreamReader and fill a Writer with its JSON representation
+     * by calling com.fasterxml.jackson write functions according to input type.
+     *
+     * Implements basic part of the specification. String (un)escaping is fully delegated to jackson
+     * and NOT fully conforming to spec.
+     *
+     * @param nodeValue the NodeValue to be read
+     * @param writer the Writer to be used
+     * @throws XPathException on error in XML JSON input according to specification
+     */
+    private void nodeValueToJson(NodeValue nodeValue, Writer writer) throws XPathException {
+        StringBuilder tempStringBuilder = new StringBuilder();
+        JsonFactory jsonFactory = new JsonFactory();
+        //use Stack<Object> to store String type keys and non-string type separators
+        final Integer stackSeperator = 0;
+        Stack<Object> mapkeyStack = new Stack<Object>();
```

Comment:
You probably want an `ArrayDeque` and not a `Stack`

Reply:
Thanks for the suggestion, I will look into this.

Reply:
`ArrayDeque` cannot return an elements position, can it?
I need to know if the current element already exists _before_ the 'separator'-element.
`contains()` does not yield sufficient information.
I don't see how this can be implemented by `ArrayDeque` (without getting an Iterator and doing the position search by hand).

Reply:
You would need to write your own `indexOf` for ArrayDeque. However `Stack` imposes a synchronisation overhead on every function call as it is thread-safe, which you don't need here. Maybe neither are the ideal structure for you...

Reply:
Now using `ArrayList` which should not have the synchronisation overhead, but provides an `indexOf`.